### ISSUE TITLE
strange dispatcher / async problem

### ISSUE THREAD

**Issue by IanBelchamber** [windows-toolkit/Microsoft.Toolkit.Win32#224]
I have some code which works fine in UWP but not in Xaml Islands. I simplified it to this simple example. I have a button in my top level user control. Here is the button click handler:

 ```
int count = 0;

        private async void bclick(object sender, RoutedEventArgs e)
        {
            count += 1;
            await Dispatcher.TryRunAsync(CoreDispatcherPriority.Normal, () =>
            //await Task.Run(() =>
            {
                var s = simple();
            });

            count -= 1;
        }

        private int simple()
        {
            int a = 1;
            return a;
        }
```

Every time I click the button, "var s = simple();" is executed. But the await does not return. So on each click, count increments and never decrements. 

Using "await Task.Run" instead works as I would expect. This does not look like normal deadlock as the await code is actually running, but the await is not releasing.

What is happening here and what is the solution?

Thanks

**Comment by marb2000** [comment]
CoreDispatcher is not reliable with win32 given it was designed for UWP only. XAML Islands should use DispatcherQueue which is using CoreMessaging that works with win32 message-pumps.

**Comment by marb2000** [comment]
I believe that Dispatcher.TryRunAsync  is not working on the context WPF and UWP. Short story: There are some type converters from UWP to .NET and vice-versa in the WindowsRuntime dlls that contains a bug.
There are two workarounds:

**A) Force to use the .NET Tasks type using Task.Run()**

```
private async void bclick(object sender, RoutedEventArgs e)
   {
       count += 1;
await Task.Run(() => Dispatcher.TryRunAsync(CoreDispatcherPriority.Normal, () =>
             {
                 var s = simple();
             }));
count -= 1;
       }

```

**B) Use DispatcherQueue**
```
       private async void bclick(object sender, RoutedEventArgs e)
        {
            count += 1;
            TaskCompletionSource<int> tcs = new TaskCompletionSource<int>();

            bool isQueued = DispatcherQueue.GetForCurrentThread().TryEnqueue(
                    () =>
                    {
                        try
                        {
                            tcs.TrySetResult(simple());
                        }
                        catch (Exception ex)
                        {
                            tcs.TrySetException(ex);
                        }
                    });

            int resultFromSimpleCallOnOtherThread = await tcs.Task;

            count -= 1;
        }
```

**Comment by IanBelchamber** [comment]
Hi, many thanks for the response. Maybe there are some differences on my system, but I already tried something like A). I believe that this is not awaiting the dispatcher, so I found that although "async void bclick" completed normally, I had a resource leak as the task itself was still not completing. 

I also tried B), it works if the dispatcher thread calls bclick, but if it is not the dispatcher, GetForCurrentThread returns null. In my case, I have some code that updates GUI controls which could be called from the dispatcher or not, I want it to work the same.

The solution I have ended up with is to use B), but to get the Windows.System.DispatcherQueue from the window on the WPF side, and use it on the UWP side, this seems to work correctly without resource leaks.

Many thanks, Ian
